src/NovelRT/Audio/AudioService.cpp

```diff
@@ -3,363 +3,349 @@
-AudioService::AudioService() : _nextChannel(1), _musicTime(0), _musicPausedTime(0), isInitialized(false), _logger(Utilities::Misc::CONSOLE_LOG_AUDIO) {
-  initializeAudio();
+AudioService::AudioService() : _device(Utilities::Lazy<std::unique_ptr<ALCdevice, void(*)(ALCdevice*)>> (std::function<ALCdevice*()>([this] {
+    auto device = alcOpenDevice((_deviceName.empty())? nullptr : _deviceName.c_str());
+    if (!device) {
+      throw std::runtime_error("OpenAL: Could not get audio devices!");
+    }
+    return device;
+  }), [](auto x) { alcCloseDevice(x); })),
+  _context(Utilities::Lazy<std::unique_ptr<ALCcontext, void(*)(ALCcontext*)>>(std::function<ALCcontext*()>([this] {
+    auto context = alcCreateContext(_device.getActual(), nullptr);
+    alcMakeContextCurrent(context);
+    isInitialised = true;
+    _deviceName = alcGetString(_device.getActual(), ALC_DEVICE_SPECIFIER);
+    _logger.logInfo("OpenAL Initialized on device: ", _deviceName);
+    return context;
+  }), [](auto x) {
+    alcMakeContextCurrent(nullptr);
+    alcDestroyContext(x);
+  })), isInitialised(false), _logger(Utilities::Misc::CONSOLE_LOG_AUDIO),
+    _musicSource(), _musicSourceState(0), _soundSource(), _soundSourceState(0) {
 }
 
 bool AudioService::initializeAudio() {
-  logIfSDLFailure(SDL_InitSubSystem, SDL_INIT_AUDIO);
-  logIfMixerFailure(Mix_OpenAudio, 44100, MIX_DEFAULT_FORMAT, 2, 2048);
-  logIfMixerFailure(Mix_AllocateChannels, (Uint32)NOVEL_MIXER_CHANNELS);
-  _logger.logInfoLine("SDL2_Mixer Initialized.");
-  isInitialized = true;
-  return isInitialized;
+  _device.getActual();
+  _context.getActual();
+  alGenSources(1, &_musicSource);
+  alGenSources(1, &_soundSource);
+  alSourcef(_musicSource, AL_GAIN, 0.75f);
+  alSourcef(_soundSource, AL_GAIN, 0.75f);
+  alSourcef(_musicSource, AL_PITCH, _pitch);
+  alSourcef(_soundSource, AL_PITCH, _pitch);
+
+  return isInitialised;
 }
 
-void AudioService::load(std::string input, bool isMusic) {
-  if (!isMusic)
-  {
-    auto exists = _sounds.find(input);
-    if (exists != _sounds.end()) return;
+ALuint AudioService::readFile(std::string input) {
+  SF_INFO info;
+  info.format = 0;
+  SNDFILE* file = sf_open(input.c_str(), SFM_READ, &info);
 
-    Mix_Chunk* newSound = Mix_LoadWAV(input.c_str());
-    if (newSound != nullptr)
-    {
-      _sounds[input] = newSound;
-    }
-    else
-    {
-      _logger.logError("SDL_Mixer error occurred during load. Error: ", getSDLError());
-    }
+  if (file == nullptr) {
+    _logger.logErrorLine(std::string(sf_strerror(nullptr)));
   }
-  else
-  {
-    auto exists = _music.find(input);
-    if (exists != _music.end()) return;
 
-    Mix_Music* newMusic = Mix_LoadMUS(input.c_str());
-    if (newMusic != nullptr)
-    {
-      _music[input] = newMusic;
-    }
-    else
-    {
-      _logger.logError("SDL_Mixer error occurred during load. Error: ", getSDLError());
-    }
-  }
-}
+  std::vector<uint16_t> data;
+  std::vector<short> readBuffer;
+  readBuffer.resize(4096);
```

Comment:
Why 4096? Also the default buffer size should probably be a defined constant.

Reply:
This was changed to 2048 and put behind a const.
4096 was a random guess, but 2048 seems to be a happy medium between fast loading and increased memory usage - for the new audio API, we would want to discern between files needing a larger amount of buffer space (meaning a larger buffer size), and files needing a smaller amount (like sfx, which would need a smaller buffer size so that we don't allocate useless space).